src/Xamarin.Android.Build.Tasks/Tasks/ResolveLibraryProjectImports.cs

```diff
@@ -154,26 +160,95 @@ static string GetTargetAssembly (ITaskItem assemblyName)
 					.Select (a => GetTargetAssembly (a))
 					.Where (a => a != null)
 					.Distinct ()) {
-				foreach (var imp in new string [] {imports_dir, "library_project_imports"}.Distinct ()) {
-					string assemblyIdentName = Path.GetFileNameWithoutExtension (assemblyPath);
-					if (UseShortFileNames) {
-						assemblyIdentName = Xamarin.Android.Tasks.MonoAndroidHelper.GetLibraryImportDirectoryNameForAssembly (assemblyIdentName);
-					}
-					string outDirForDll = Path.Combine (OutputImportDirectory, assemblyIdentName);
-					string importsDir = Path.Combine (outDirForDll, imp);
+				string assemblyIdentName = Path.GetFileNameWithoutExtension (assemblyPath);
+				if (UseShortFileNames) {
+					assemblyIdentName = Xamarin.Android.Tasks.MonoAndroidHelper.GetLibraryImportDirectoryNameForAssembly (assemblyIdentName);
+				}
+				string outDirForDll = Path.Combine (OutputImportDirectory, assemblyIdentName);
+				string importsDir = Path.Combine (outDirForDll, ImportsDirectory);
+#if SEPARATE_CRUNCH
+				// FIXME: review these binResDir thing and enable this. Eclipse does this.
+				// Enabling these blindly causes build failure on ActionBarSherlock.
+				//string binResDir = Path.Combine (importsDir, "bin", "res");
+				//string binAssemblyDir = Path.Combine (importsDir, "bin", "assets");
+#endif
+				string resDir = Path.Combine (importsDir, "res");
+				string assemblyDir = Path.Combine (importsDir, "assets");
+
+				// Skip already-extracted resources.
+				var stamp = new FileInfo (Path.Combine (outdir.FullName, assemblyIdentName + ".stamp"));
+				if (stamp.Exists && stamp.LastWriteTime > new FileInfo (assemblyPath).LastWriteTime) {
+					Log.LogDebugMessage ("Skipped resource lookup for {0}: extracted files are up to date", assemblyPath);
 #if SEPARATE_CRUNCH
-					// FIXME: review these binResDir thing and enable this. Eclipse does this.
+					// FIXME: review these binResDir/binAssemblyDir thing and enable this. Eclipse does this.
 					// Enabling these blindly causes build failure on ActionBarSherlock.
-					//string binResDir = Path.Combine (importsDir, "bin", "res");
-					//string binAssemblyDir = Path.Combine (importsDir, "bin", "assets");
+					if (Directory.Exists (binResDir))
+						resolvedResourceDirectories.Add (binResDir);
+					if (Directory.Exists (binAssemblyDir))
+						resolvedAssetDirectories.Add (binAssemblyDir);
 #endif
-					string resDir = Path.Combine (importsDir, "res");
-					string assemblyDir = Path.Combine (importsDir, "assets");
+					if (Directory.Exists (resDir))
+						resolvedResourceDirectories.Add (resDir);
+					if (Directory.Exists (assemblyDir))
+						resolvedAssetDirectories.Add (assemblyDir);
+					continue;
+				}
+
+				if (Directory.Exists (outDirForDll))
+					Directory.Delete (outDirForDll, true);
```

Comment:
The Hash has been replaced by a lookup cache file which will map the assembly to a integer index directory. So you get something like

MyProject=1
Xamarin.Android.Support.v7.AppCompat.21.0.3.0=2

So on the path we only take up "1" or "2" (up to the number of dependencies). Rather than a ton of redundant characters.